SALP: Add interface toggle to enable/disable sites dashboard as landing page

vykes-mac opened the pull request:
#### Proposed Changes

* Add a component that toggles user's preference for setting sites dashboard as landing page.

#### Testing Instructions

- Go to `/me/account`
- Verify sites as landing page toggle is shown when  config `'sites-as-landing-page': true` is set
- Verify sites as landing page toggle is not shown when  config `'sites-as-landing-page': false` 
- Verify event `calypso_settings_sites_dashboard_landing_page_toggle ` event is sent in console when toggle on/off
- Run `localStorage.setItem( 'debug', 'calypso:analytics*' );` in console to see events.
<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* 
![image](https://user-images.githubusercontent.com/47489215/203648986-6c735db2-3a52-4498-825c-e523d26f9d09.png)


#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #

vykes-mac committed:
Add component to toggle sites as dashboard interface preference

vykes-mac committed:
update preference to be same name from #70356

p-jackson commented on client/me/account/main.jsx:
```diff
@@ -968,6 +976,16 @@ class Account extends Component {
+						{ config.isEnabled( 'sites-as-landing-page' ) && (
+							<FormFieldset>
+								<FormLabel id="account__default_landing_page">
+									{ translate( 'Sites Dashboard as landing page' ) }
```
We've tried not to call it a dashboard: p1659127658803489/1659090371.853969-slack-C0347E545HR

The annouceent post just called it the "Sites page"
https://wordpress.com/blog/2022/09/01/introducing-a-new-way-to-access-all-your-sites/

"Sites as landing page" isn't the clearest, but it is only a heading. The toggle label can be more descriptive I guess.

vykes-mac replied:
Oh yeah, forgot we are steering clear of the term "Dashboard". for now I'll use "Sites as landing page" for heading and for the description "Set sites as the default landing page".

p-jackson commented on client/me/account/main.jsx:
```diff
@@ -217,6 +219,12 @@ class Account extends Component {
 		this.setState( { userLoginConfirm: event.target.value } );
 	};
 
+	updateSitesAsLandingPage = ( checked ) => {
```
Probably just personal taste, but we could move the tracks handling into `<ToggleSitesAsLandingPage>` component. The parent doesn't really need to be involved in what the child component is doing in this case.

vykes-mac replied:
Can be moved there to encapsulate everything, I was looking at some previous implementation  but It might be done that way for testing purposes so there's less dependencies to mock if we write test for the component.

p-jackson commented on client/me/account/toggle-sites-as-landing-page.tsx:
```diff
@@ -0,0 +1,53 @@
+function ToggleSitesAsLandingPage( { onToggle, className }: ToggleSitesLandingPageProps ) {
+	const translate = useTranslate();
+	const dispatch = useDispatch();
+	const useSitesAsLandingPage = useSelector( ( state ) =>
+		getPreference( state, 'sites-landing-page' )
+	)?.useSitesAsLandingPage;
```
It's usually better to select the data you need _inside_ the `useSelector` callback. That way it only re-renders when the specific preference you're interested in changes, not when any preference changes.

```suggestion
	const useSitesAsLandingPage = useSelector( ( state ) =>
		getPreference( state, 'sites-landing-page' )?.useSitesAsLandingPage
	);
```

More here under "Select the data you need inside the callback": p7H4VZ-2Q6-p2 (this P2 is talking about GB's `useSelect` specifically, but the same principle applies).

vykes-mac replied:
TIL: thank you for that, I was actually following your [PR](https://github.com/Automattic/wp-calypso/pull/70356) but missed the fact that it was on the `getPreference`  method and not on the selector 😅

p-jackson replied:
Yeah this particular one is sorta easy to miss! Bracket-blindness or something like that 😅

p-jackson commented on client/me/account/toggle-sites-as-landing-page.tsx:
```diff
@@ -0,0 +1,53 @@
+				duration: 10000,
+			} )
+		);
+		setIsDisabled( false );
```
There's a small memory leak here. If you click the toggle and then very quickly navigate away, then by the time the network request comes back and this async function resumes the component will have unmounted. You'll see the somewhat familiar React error message `setIsDisabled` is called.

<img width="676" alt="CleanShot 2022-11-25 at 13 32 12@2x" src="https://user-images.githubusercontent.com/1500769/203878187-2516f315-6217-4c52-b0d0-0fc6ae50dfcb.png">

I'm not sure of a quick way to tidy this up. One way would be to add a `isSaving()` selector to the preferences store, that way we wouldn't need to store the `isDisabled` state in our component.
I think there might also be a way to do this by wrapping our saving logic in ReactQuery. I'll have more of a think about how that can be done, just for fun 😄 

It's unlikely the memory leak will amount to much. Users probably aren't spending all their time in Calypso toggling this setting. So I don't think it's a show-stopper.

p-jackson replied:
Here's an idea for how the leak could be fixed using ReactQuery
https://github.com/Automattic/wp-calypso/compare/add/sites-landing-page-toggle...add/sites-landing-page-toggle-react-query

The idea is that ReactQuery doesn't just have to be used to initiate network requests, it's abstractions can be used to help with anything async.

The other option I mentioned was to move the "is saving" state out of the component and into Redux. Both of these options prevent the leak in the same way—moving the state out of the React tree and thus aren't affected by components being mounted/unmounted.

vykes-mac replied:
Thank you for the info on react-query, that's pretty interesting how it works in this context. Always thought it was for network requests only haha.  

I think we can use the `savePreference` as is without a disabled state since it calls `setPreference` action to save to local state before issuing the call to store it remotely, with this I do not block the UI and memory leak issue is solved as I'm not using `useState` hook no more.

p-jackson replied:
Yes removing the disabled state altogether also fixes it. Sometimes the disabled state is there to fix weirdness when you rapidly toggle the setting though. Like if you re-toggle it while a network request is still underway it can do weird things. But if that's not the case then it doesn't matter and we can just remove the disabled state.

p-jackson replied:
Yes like this:

https://user-images.githubusercontent.com/1500769/204163752-ea4d8920-0179-4f9d-91ff-f457c180ef91.mov

If you toggle it on and then off again quickly (e.g. if the user misclicked or they changed their mind) then when the network requests come back they cause the control to toggle. This is why we tend to disable these toggle buttons until we've finished saving.

p-jackson replied:
To be fair this isn't the only place in the UI that has this bug ...


https://user-images.githubusercontent.com/1500769/204163930-141505c1-3f88-4fa0-9cf9-6b3a8f9ba718.mov

vykes-mac replied:
yeah I realise that issue and It's present alot of places in the codebase as you mentioned.  I've added the a `isSavingPreference` selector that is used to disable the button. I think it might be more helpful with the selector since it can be reused, that using the react-query way.

p-jackson reviewed (commented):
The e2e tests should be fixed after a rebase. See p1669325844642229-slack-C02DQP0FP

p-jackson commented on client/me/account/main.jsx:
```diff
@@ -968,6 +976,16 @@ class Account extends Component {
+							<FormFieldset>
+								<FormLabel id="account__default_landing_page">
+									{ translate( 'Sites Dashboard as landing page' ) }
+									<Badge type="info-blue">beta</Badge>
```
Also I think we need something like a `margin-left: 8px` on this badge

vykes-mac committed:
refactor to prevent memory leak

vykes-mac committed:
Merge branch 'trunk' into add/sites-landing-page-toggle

vykes-mac commented on client/me/account/main.jsx:
```diff
@@ -87,6 +90,9 @@ const INTERFACE_FIELDS = [
 	'calypso_preferences',
 ];
 
+const BetaBadge = styled( Badge )`
```
didn't feel right creating this styled component just for margins but inline styling of the `<Badge/>` component doesn't seem to work. maybe the component itself needed to modified to accept style props.

p-jackson replied:
The inline style wouldn't work because `<Badge>` doesn't forward any `style` prop through to its children. However it _does_ forward the `className` prop, which is why Emotion works.

The accounts page doesn't use Emotion, so I'd be reluctant to add it just for this one margin. If you take a look at the existing code in the `<Account>` component you'll see that they're applying CSS classes—which have been defined in `client/me/account/style.scss`—directly to elements.

So the thing to do here would be to define a new `account__beta-badge` class in `style.scss` and then apply that class directly to the `<Badge>` component in the render method. No need to define a new component.

vykes-mac replied:
yeah defining the class make sense. Including emotion just didn't feels right.

vykes-mac committed:
dispatch recordTracksEvent

vykes-mac committed:
remove duplicate save preference call

p-jackson reviewed (changes requested):
Let's not introduce Emotion to these components which already use scss for styling.

vykes-mac committed:
add css class to style data badge

vykes-mac committed:
add selector to determine when preference is being saved.

zaguiini commented on client/me/account/main.jsx:
```diff
@@ -968,6 +970,18 @@ class Account extends Component {
+								<FormLabel id="account__default_landing_page">
+									{ translate( 'Sites as landing page' ) }
+									<Badge className="account__beta-badge" type="info-blue">
+										beta
```
We need to translate this literal:

```suggestion
										translate( 'beta' )
```

zaguiini commented on client/me/account/toggle-sites-as-landing-page.tsx:
```diff
@@ -0,0 +1,53 @@
+
+	async function handleToggle( isChecked: boolean ) {
+		const preference = { useSitesAsLandingPage: isChecked, updatedAt: Date.now() };
+		dispatch(
+			recordTracksEvent( 'calypso_sites_as_landing_page_toggled', {
+				sites_as_landing_page: isChecked,
+			} )
+		);
```
Can we dispatch this event only after saving the preference?

zaguiini commented on client/me/account/toggle-sites-as-landing-page.tsx:
```diff
@@ -0,0 +1,53 @@
+				checked={ !! useSitesAsLandingPage }
+				onChange={ handleToggle }
+				disabled={ isSaving }
+				label={ translate( 'Set the sites management page as the default landing page.' ) }
```
I'd say something like: "Make the Sites page (with a link to it) my landing page" reads better. Thoughts?

p-jackson replied:
The link would be worth a thousand words for sure

zaguiini commented on client/me/account/toggle-sites-as-landing-page.tsx:
```diff
@@ -0,0 +1,53 @@
+		await dispatch( savePreference( 'sites-landing-page', preference ) );
+
+		dispatch(
+			successNotice( translate( 'Settings saved successfully!' ), {
```
Do we need this notice? I think the fact that the toggle is now active is enough, visually, to indicate that the preference was saved 🤔

p-jackson replied:
Doesn't seem super necessary, but it's consistent with the other interface settings at least. One benefit might be to avoid confusion for users who are looking for a save button. Some settings have a save button and some don't, so maybe it's good settings that update immediately to have this notice to prevent users looking for something that doesn't exist.

p-jackson replied:
I don't know if that's the real reason, just my posthoc rationalisation 😅 I'm not that attached to the notice.

vykes-mac replied:
I'm of the save view here as @p-jackson, since there's no save button in this UI I assumed the notification is to indicate that the action was saved. The other components here shows the notice so I keep it the same.

zaguiini replied:
Fair enough, let's keep it!

zaguiini reviewed (commented):
Looking good so far. A few suggestions!

p-jackson commented on client/state/preferences/reducer.js:
```diff
@@ -72,6 +72,16 @@ export const fetching = ( state = false, action ) => {
+	switch ( action.type ) {
+		case PREFERENCES_SET:
+			return true;
+		case PREFERENCES_SAVE_SUCCESS:
+			return false;
```
We should handle the network failure case too to make sure the boolean doesn't get stuck `true` forever.

```suggestion
		case PREFERENCES_SAVE_SUCCESS:
		case PREFERENCES_SAVE_FAILURE:
			return false;
```

Take a look at the `PREFERENCES_FETCH_FAILURE` action to see where it's dispatched.

p-jackson commented on client/me/account/toggle-sites-as-landing-page.tsx:
```diff
@@ -0,0 +1,53 @@
+import { getPreference, isSavingPreference } from 'calypso/state/preferences/selectors';
+
+type ToggleSitesLandingPageProps = {
+	onToggle?: ( checked: boolean ) => void;
+	className?: string;
```
We don't use either of these props so we could just remove them 🤷‍♂️ No biggy though.

p-jackson commented on client/me/account/toggle-sites-as-landing-page.tsx:
```diff
@@ -0,0 +1,53 @@
+	const useSitesAsLandingPage = useSelector(
+		( state ) => getPreference( state, 'sites-landing-page' )?.useSitesAsLandingPage
+	);
+	const isSaving = useSelector( ( state ) => isSavingPreference( state ) );
```
Great! This redux method is probably more familiar to everyone.

p-jackson reviewed (approved):
Change is looking good to me 👍 

I agree with Luis' comments, and think we should handle the network error case (see below), but after that, I think it's good to go.

p-jackson commented on client/me/account/toggle-sites-as-landing-page.tsx:
```diff
@@ -0,0 +1,53 @@
+			<ToggleControl
+				checked={ !! useSitesAsLandingPage }
+				onChange={ handleToggle }
+				disabled={ isSaving }
```
The build is failing here because the `@wordpress/components` type definitions don't include the `disabled` props 😭 This is always frustrating.

Here's a hack that doesn't require `@wordpress/components` to update before we can merge:
```typescript
import { ToggleControl as OriginalToggleControl } from '@wordpress/components';

const ToggleControl = OriginalToggleControl as React.ComponentType<
	OriginalToggleControl.Props & {
		disabled?: boolean;
	}
>;
```

vykes-mac committed:
add action to cover failures when saving preferences

vykes-mac committed:
refactor to remove unused props and add a link to sites page in label description.

p-jackson commented on client/me/account/toggle-sites-as-landing-page.tsx:
```diff
@@ -0,0 +1,62 @@
+		);
+
+		dispatch(
+			recordTracksEvent( 'calypso_sites_as_landing_page_toggled', {
```
I'm thinking maybe we should change the event names for consistency: https://github.com/Automattic/wp-calypso/pull/70512

So maybe this should be

```suggestion
			recordTracksEvent( 'calypso_sites_dashboard_landing_page_setting_toggle', {
```

Or maybe `_click`?

It might depend on the answer to this p1669699187812099-slack-C0347E545HR

zaguiini replied:
I'd say `calypso_settings_sites_dashboard_landing_page_toggle`. It's good to understand the origin of the event -- it's almost always the prefix.

zaguiini commented on client/me/account/toggle-sites-as-landing-page.tsx:
```diff
@@ -0,0 +1,62 @@
+	const label = createInterpolateElement(
+		translate( 'Make the <a>Sites page</a> my default landing page.' ),
+		{
+			a: <a href={ localizeUrl( '/sites' ) } />,
```
Let's open this in a new tab:

```suggestion
			a: <a target="_blank" href={ localizeUrl( '/sites' ) } />,
```

vykes-mac committed:
change event name and open link in a new tab

p-jackson committed:
Remove unnecessary localizeUrl()

localizeUrl() is only necessary when linking to a page where the
content's language depends on the URL e.g. a support site.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7645967

Thank you @vykes-mac for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7645967).
